sssd_test_framework/utils/virtual_smartcard.py

```diff
@@ -0,0 +1,205 @@
+        :type cert_path: str, optional
+        :param subj: Subject for the certificate, defaults to "/CN=Test Cert"
+        :type subj: str, optional
+        :return: Tuple of (key_path, cert_path)
+        :rtype: tuple
+        """
+        self.host.conn.exec(
+            [
+                "openssl",
+                "req",
+                "-x509",
+                "-nodes",
+                "-sha256",
+                "-days",
+                "365",
+                "-newkey",
+                "rsa:2048",
+                "-keyout",
+                key_path,
+                "-out",
+                cert_path,
+                "-subj",
+                subj,
+            ]
+        )
+        return key_path, cert_path
+
+    def insert_card(self) -> None:
+        """
+        Simulates card insertion by starting the smart card service.
+        """
+        self.host.conn.exec(["systemctl", "start", "virt_cacard.service"])
+
+    def remove_card(self) -> None:
+        """
+        Simulates card removal by stopping the smart card service.
+        """
+        self.host.conn.exec(["systemctl", "stop", "virt_cacard.service"])
+
+    def reset_service(self) -> None:
+        """
+        Restarts the virtual smart card service.
+        """
+        self.host.conn.exec(["systemctl", "restart", "virt_cacard.service"])
+
+    def setup_local_card(self, client: Client, username: str) -> None:
+        """
+        Setup local system for smart card authentication
+
+        .. code-block:: python
+            :caption: Example usage
+
+            @pytest.mark.topology(KnownTopology.Client)
+            def test_example(client: Client):
+                client.smart_card.setup_local_card(client, 'localuser1')
+
+                result = client.host.conn.run("su - localuser1 -c 'su - localuser1 -c whoami'", input="123456")
+                assert "PIN" in result.stderr
+                assert "localuser1" in result.stdout
+
+        :param client: client object needed for sssd and other util configurations.
+        :type client: Client
+        :param username: Name of user to create and configure for local smart card authentication.
+        :type username: str
+        """
+        (key, cert) = self.generate_self_signed_cert()
+
+        self.init()
+        self.add_key(key)
+        self.add_cert(cert)
+        self.reset_service()
+
+        client.sssd.sssd["domains"] = "local"
+        client.sssd.common.local()
+        client.sssd.dom("local")["local_auth_policy"] = "only"
+        client.sssd.section(f"certmap/local/{username}")["matchrule"] = "<SUBJECT>.*CN=Test Cert*"
+        client.sssd.pam["pam_cert_auth"] = "True"
+        client.host.conn.run("echo >> /etc/sssd/pki/sssd_auth_ca_db.pem")
+        client.host.conn.run(f"cat {cert} >> /etc/sssd/pki/sssd_auth_ca_db.pem")
+        client.sssd.start()
+
+        client.local.user(f"{username}").add()
```

Comment:
I think this should not be included because you can set up a card and use a local certificate instead of a local account. So, this should remain a setup step.

Reply:
You're right.    That should move into the test because it would allow us to cover different scenarios.   Also, that reminds me that the matchrule above should use a parameter too.   I'll add a comment about that too.